### ISSUE TITLE
Attribute format and range of values

### ISSUE THREAD

**Issue by Christian-Shom** [iho-ohi/S-101-Documentation-and-FC#104]
Although S-57 attribute catalogue defines a format for type Float that does not always fulfil the need for some features (ex : xxx.x for RADIUS), it is possible to use larger values (ex: 1500 for RADIUS). S-58 check 26b contains the following guidance: "_for attribute values of type "float", the resolution given  in the format statement by  the integer part (for  example XX.X) must not be checked_" and Validation software do not trigger errors in such cases.
- Is this going to work similarly in S-101, now that we have xml FCs?
- should we include some introductory Notes at the start of Sections 27 and 28 of the DCEG stating that the Format indication is  indicative only and no padded zeros should be used?
- should we consider using the upper and lower permitted values within the Feature Catalogue? (this would avoid having validation checks such S-58 26a and 26b in S-101).
- should we simply add an additional "x" to the format descriptions with a statement to not include padded zeros?

**Comment by rmalyankar** [comment]
Validation checks should certainly use the feature catalogue, specifically the _constraints_ element in S100_FC_SimpleAttribute, if and when that element is populated (and it should be populated for all numeric attributes if at all possible).

The format as described in the DCEG should conform to the XML feature catalogue and be treated only as human-readable guidance to the encoder (i.e., in case of conflict the FC should prevail). In a few cases the DCEG might supplement what is in the FC with additional constraints which cannot be encoded using the _constraints_ element.

DCEG 2.4.2 includes a statement about not using non-significant zeros to pad, along with examples. This should be extended with a statement about omitting the portion from the decimal point and subsequent zeros if the decimal point is followed by only zeros (cf. S-100 10a-5.1.4).

Scientific convention is that zeros after a decimal point _are_ significant even if there are only zeros, so if S-101 wants to omit them it should explicitly say so.

**Comment by Christian-Shom** [comment]
There are only 6 constraints in the entire current FC, the most complete being for depthRangeMaximumValue:

![image](https://github.com/iho-ohi/S-101-Documentation-and-FC/assets/71139640/f6ea8308-a3fe-45d2-a252-abd2745cd05c)

This is far from reflecting the entire DCEG. 

- Should we review it to make it the reference (and not the DCEG)?
- I do not understand the constraints on attribute **source** in the FC:
![image](https://github.com/iho-ohi/S-101-Documentation-and-FC/assets/71139640/473ae34c-9470-4184-bab3-b90f0edb46ce)

**Comment by FrankHippmann** [comment]
I agree with encoding the constraints in the FC, and having lower and upper bounds where appropriate. Other fields that are encoded in the DCEG could also be transferred to the FC:

- Unit
- Resolution

Some attribute values allow multiple patterns, e.g., reportedDate

- YYYYMMDD (full date, mandatory)
- YYYYMM-- (no specific day required – mandatory)
- YYYY---- (no specific month required – mandatory)

Regarding zeros after the decimal place, I agree with Raphael that scientifically, 10 and 10.00 have different meanings, with 10.00 having higher precision, and the 0s following the decimal point **are** significant. The statement in 10a-5.1.4 implies that if there are only 0s after the decimal point, they are not significant. This is false. However, whether 0s after the decimal point have any practical use remains to be seen, as there is already metadata indicating uncertainty of values, and in S-57 they have been ignored.

**Comment by Christian-Shom** [comment]
> Other fields that are encoded in the DCEG could also be transferred to the FC:
> 
>     * Unit
> 
>     * Resolution

I agree we should try to have more consistency between S-100, S-101 FC and S-101 DCEG.
I suggest populating in the FC, everywhere needed, S-100 items:

- Attribute type (Ex: Real)
- Attribute constraints:
     - precision
     - stringLength
     - textPattern
     - Range:
          - rangeLower
          - rangeUpper
          - rangeClosure

In the DCEG (section 27) we have:
- Attribute type - **OK**;
- Unit - **Should be transferred to the FC**;
- Resolution - **OK**, refer to S-100 precision but is easier to understand for encoders;
- Format - Seems redundant with Attribute type + Resolution - **I suggest taking it off**;
- Minimum value - **OK** refer to S-100 rangeLower;
- Maximum value - **OK** refer to S-100 rangeUpper.

**Comment by Richard-Fowle** [comment]
I agree with Christian's comments and would add that we also need to ensure that there is consistency between the DCEG and the Data Dictionary Register (for example DCEG 27.75 Indication states 'Each Channel should be indicated in square brackets by 4 digits and up to 4 characters (A-Z).' this would be up to 10 characters,  but the Data Dictionary Register states that Length  'Up to 6 characters'

**Comment by JeffWootton** [comment]
Note paper [S-101PT12-06.12](https://iho.int/uploads/user/Services%20and%20Standards/S-100WG/S-101PT12/S-101PT12_2024_06.12_EN_Attribute_Format_and_Constraints_V1.pdf) and accompanying [presentation](https://iho.int/uploads/user/Services%20and%20Standards/S-100WG/S-101PT12/S-101PT12_2024_06.12A_EN_Attribute_Format_Constraints_Preso.pdf). See also recommendation included in paper [S-101PT12-05.1.1](https://iho.int/uploads/user/Services%20and%20Standards/S-100WG/S-101PT12/S-101PT12_2024_05.1.1_EN_GitHub_Open_Issues_Impacting_on_the_FC_V1.pdf): _Note paper S-101PT12-06.12. Move forward based on discussions. Not critical for Edition 1.2.0. Address in slower time._

Recommendation approved at S-101PT12. Issue to remain open.

**Comment by DavidGrant-NIWC** [comment]
There is an additional ambiguity regarding the precision: is it the number of significant digits, or the number of digits following the decimal point? S-100 merely describes it as:
![image](https://github.com/iho-ohi/S-101-Documentation-and-FC/assets/62906211/f6e1ab4a-b175-44c9-ba04-55898b7fe1ed)

The presentation implies the second definition:
![image](https://github.com/iho-ohi/S-101-Documentation-and-FC/assets/62906211/596451bc-0cb4-4851-a3e9-7502d4066199)


https://simple.wikipedia.org/wiki/Arithmetic_precision

**Comment by JeffWootton** [comment]
Attribute formats and range values have been amended throughout (DCEG and GI Registry) to conform with the conventions described in S-100.

For precisions, the convention of expressing the exponent of Real type attributes (0.1, 0.01, ...) has been retained in the DCEG, however the following guidance has been included at clause 2.4.2:

![image](https://github.com/iho-ohi/S-101-Documentation-and-FC/assets/70211773/702a13b1-378c-446b-8027-42fbbebe23e9)

For range constraints (numeric and character) estimates have been made based on research with an added margin intended to allow for complete inclusivity. These constraints may be emended in future based on encoding experience.

Consider that this Issue can be closed and new Issues opened as appropriate.

**Comment by DavidGrant-NIWC** [comment]
> For precisions, the convention of expressing the exponent of Real type attributes (0.1, 0.01, ...) has been retained in the DCEG, however the following guidance has been included at clause 2.4.2:
> 
> ![image](https://private-user-images.githubusercontent.com/70211773/331089954-702a13b1-378c-446b-8027-42fbbebe23e9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NjY5ODAsIm5iZiI6MTcxNTg2NjY4MCwicGF0aCI6Ii83MDIxMTc3My8zMzEwODk5NTQtNzAyYTEzYjEtMzc4Yy00NDZiLTgwMjctNDJmYmJlYmUyM2U5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDEzMzgwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyNWU2NDNlZmIzMjNhMzBkMjY2MzhhYmNjZmFiZmFkMjA0ODc0ZGQ1MDYyZTE0ZjAwZGU5ZThhYWFiMjZmYzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.44ACFhOd0KdZCqSNa1fk_EbuIl8XcLZIW6UJRxFF-IQ)

This seems to be equivalent to the XML constraint _fractionDigits_, you may want to adopt a similar definition: https://www.w3.org/TR/xmlschema-2/#rf-fractionDigits

The interpretation should be clarified within S-100 so that other products do not implement conflicting interpretations; consideration should be given to renaming the constraint (_fractionDigits_ seems appropriate). However, those issues are out of scope for this group.

**Comment by JeffWootton** [comment]
There is still some confusion regarding the Precision field. Decision from S-101PT13 was to review in slower time (post S-101 Edition 2.0.0). In addition, RegEx needs to be defined for standardized text strings.

Value ranges (lower, upper and rangeClosure) for numeric attributes have been defined in the DCEG and GI Registry and reflected in the S-101 Edition 1.4.1 Feature Catalogue; and formats for numeric attributes removed.

Issue to remain open pending resolution of Precision and RegEx.

**Comment by TomRichardson6** [comment]
A paper has been submitted to TSM11 in March 2025 to clarify the description for precision at the S-100 level. @JeffWootton propose that a new Issue is opened to allow the population of missing textPattern values (Regex) this is dependent on resource and appropriate skills.

**Comment by JeffWootton** [comment]
TSM11 outcome: It was agreed that the Precision quoted for real type attributes as an integer value describes the maximum number of decimals allowed for the attribute (i.e. the maximum number of digits allowable to the right of the decimal point). This reflects the already populated fields in the GI Registry, Data Dictionary Register for S-101 and the guidance included in the S-101 DCEG.

It was agreed that the RegEx expressions included in the Paper to TSM11 require further work; and accordingly a new GitHub Issue is to be opened to specifically manage the development of the RegEx and this Issue can be closed.
